### PR TITLE
Derived Stores

### PR DESCRIPTION
**Pull request by crutchcorn**
This PR implements an initial `Derived` class implementation:

```ts
const count = new Store(10);

const doubleCount = new Derived({
  deps: [count],
  fn: () => count.state * 2,
})

// .mount is required for Derived to subscribe to `count`
const _cleanupDerivedListeners = doubleCount.mount();


const effect = new Effect(() => {
  deps: [doubleCount],
  fn: () => console.log(doubleCount.state))
})

// .mount is required for Effect to subscribe to `doubleCount`
const _cleanupEffectListeners = effect.mount();

count.setState(() => 20);
```

> [!NOTE]
> This API of `mount` seems wonky for application use because, well, it is. It's intended for other libraries like TanStack Form and TanStack Table's internals.
>
> We may introduce a more application-friendly `Derived` and `Effect` API for usage in apps, but it's not clear yet what that looks like. Please feel free to [jump into the discussion here](https://github.com/TanStack/store/discussions/143) for more

# TODO

- [x] Write docs for the frameworks in Quick Start
- [x] Write docs for `Derived State` that includes `prevVal` and `prevVals`
- [x] Write framework tests for `Derived` to make sure it continues to work
    - [x] React
    - [x] Angular
    - [x] Solid
    - [x] Svelte
    - [x] Vue
- [x] Migrate `Derived`/`Effect`'s side effects to explicit `mount` and `unmount` methods
- [x] Add `previousValues` to `Derived`

# Edgecase Handling

This Derived implementation even solves the diamond problem:

"a" is the root node with "b" and "c" relying on it. Then "d" relies on both "b" and "c"

Where writes to "a" will call "d" twice (with a flicker of incorrect data known as a "Glitch") because of the dual-derived nature. Our implementation does not have this problem and therefore can be considered "glitch-free"

This is a push-based/hot signals implementation 

# Benchmarks

Due to the relatively naive nature of this code (I wrote it in one night at ~midnight) there are performance implication in using this. Namely, in our benchmarks we are a **far** cry from Angular or Solid's implementation, but come close to Vue's implementation:

### CODE REVIEW

**Comment by crutchcorn** [comment]
image

I've increased our performance and reduced our bundle size. We're no-longer the slowest on our own benchmark and are even 1.5x faster than Vue:

image

**Comment by tannerlinsley** [comment]
Woo! 🏎️

**Comment by eneajaho** [comment]
Angular is the fastest, nice 😄!

**Comment by lucasmotta** [comment]
Out of curiosity, what's the state of this PR and what's stopping it from being reviewed/merged?

File: packages/store/src/derived.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  rootStores = new Set<Store<unknown>>()
+  deps: Deps
+
+  // Functions representing the subscriptions. Call a function to cleanup
+  _subscriptions: Array<() => void> = []
+
+  // What store called the current update, if any
+  _whatStoreIsCurrentlyInUse: Store<unknown> | null = null
+
+  constructor(deps: Deps, fn: () => TState, options?: DerivedOptions<TState>) {
```
Comment (KevinVandy) [inline packages/store/src/derived.ts:22]: would it be better if this had a flat options style like `useQuery` now does?

```ts
new Derived({
  storeKey: [],
  storeFn: () => alert("expensive")
  ...options
})
```

**Comment by tannerlinsley** [comment]
I would prefer it I think

Tanner Linsley
TanStack LLC

**Comment by crutchcorn** [comment]
I was originally holding off as performance didn't seem up-to-snuff, but comparing it against the TS39 Polyfill, it seems that we're roughly double the speed of that, so I want to merge this in so we can start using ASAP

File: packages/store/src/derived.ts (hunk 1 above)
Comment (crutchcorn) [reply packages/store/src/derived.ts]: Done!

**Pull request merged**